Attempt to clean up index before remote transfer

bcully opened the pull request:
If a node crashes during recovery, it may leave temporary files behind that can consume disk space, which may be needed to complete recovery. So we attempt to clean up the index before transferring files from a recovery source. We first attempt to call the store's `cleanupAndVerify` method, which removes anything not referenced by the latest local commit. If that fails, because the local index isn't in a state to produce a local commit, then we fall back to removing temporary files by name.

Closes #104473

bcully committed:
Attempt to clean up index before remote transfer

If a node crashes during recovery, it may leave temporary files behind
that can consume disk space, which may be needed to complete recovery.
So we attempt to clean up the index before transferring files from
a recovery source. We first attempt to call the store's
`cleanupAndVerify` method, which removes anything not referenced by
the latest local commit. If that fails, because the local index isn't
in a state to produce a local commit, then we fall back to removing
temporary files by name.

Closes #104473

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

elasticsearchmachine commented:
Hi @bcully, I've created a changelog YAML for you.

bcully committed:
Update docs/changelog/115142.yaml

bcully commented:
I'm doing this as part of onboarding, and don't really have any idea what I'm doing yet. Any kind of feedback is welcome.

idegtiarenko commented on server/src/internalClusterTest/java/org/elasticsearch/recovery/TruncatedRecoveryIT.java:
```diff
@@ -104,18 +105,43 @@ public void testCancelRecoveryAndResume() throws Exception {
+                            logger.info("checking for garbage at {}", shardPath.resolve("recovery.garbage"));
+                            assertFalse(
+                                "garbage file should have been cleaned before file transmission",
+                                Files.exists(shardPath.resolve("garbage"))
```
I believe this checks for a wrong file, not the one created earlier (`recovery.garbage` vs `garbage`).
It would be nice to extract its name as variable earlier in the code and possibly randomize the file name.

bcully replied:
Whoops! Sounds good, I'll randomly generate some during round 1 and verify that they are all gone before transfer in round 2.

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into recovery-clean-before-transfer

Tim-Brooks commented on server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java:
```diff
@@ -397,6 +397,32 @@ record StartRecoveryRequestToSend(StartRecoveryRequest startRecoveryRequest, Str
+                        // everything in the directory but in that case we might want to be a little more thoughtful about
+                        // the exact exception triggered here. Corrupt indices might be recoverable without a full transfer, for
+                        // example.
+                        logger.info("cleanup before peer recovery failed on shard [{}]; exception: [{}]", indexShard.shardId(), e);
```
logger.info(() -> format("cleanup before peer recovery failed on shard [%s]", indexShard.shardId()), e);

Tim-Brooks replied:
We prefer to use the logger method which takes the exception. Also, `Supplier` is an option. But you can also pass `formatted string` directly. It's extremely common to use supplier for trace/debug logs which are normally disabled. Since warn, error, and info are normally enabled I don't think it makes a huge difference.

bcully replied:
I see. I was trying to be consistent with the [neighboring code](https://github.com/elastic/elasticsearch/blob/d6f8de22d697cd4db7a6f0eb179d6835867e9d35/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java#L364) but I can see that it is [in](https://github.com/elastic/elasticsearch/blob/d6f8de22d697cd4db7a6f0eb179d6835867e9d35/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java#L261) [flux](https://github.com/elastic/elasticsearch/blob/d6f8de22d697cd4db7a6f0eb179d6835867e9d35/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java#L256) and will use the lambda form, and not eagerly stringify the exception.

Tim-Brooks commented on server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java:
```diff
@@ -397,6 +397,32 @@ record StartRecoveryRequestToSend(StartRecoveryRequest startRecoveryRequest, Str
+                            recoveryTarget.cleanTempFiles();
+                        } catch (IOException ie) {
+                            // log and ignore. Recovery gets another chance to clean up after file transfer.
+                            logger.warn("temporary file cleanup failed on shard [{}]; exception: [{}]", indexShard.shardId(), ie);
```
logger.warn(() -> format("temporary file cleanup failed on shard [%s]", indexShard.shardId()), ie);

Tim-Brooks commented on server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java:
```diff
@@ -397,6 +397,32 @@ record StartRecoveryRequestToSend(StartRecoveryRequest startRecoveryRequest, Str
+                    Store store = indexShard.store();
+                    store.incRef();
+                    try {
+                        store.cleanupAndVerify("cleanup before peer recovery", indexShard.snapshotStoreMetadata());
```
This wraps some IOExceptions in `throw new IllegalStateException("Can't delete " + existingFile + " - cleanup failed", ex);`. Is that something we should be catching and handling?

Tim-Brooks replied:
```java
 catch (IOException ex) {
                    if (existingFile.startsWith(IndexFileNames.SEGMENTS) || existingFile.startsWith(CORRUPTED_MARKER_NAME_PREFIX)) {
                        // TODO do we need to also fail this if we can't delete the pending commit file?
                        // if one of those files can't be deleted we better fail the cleanup otherwise we might leave an old commit
                        // point around?
                        throw new IllegalStateException("Can't delete " + existingFile + " - cleanup failed", ex);
                    }
                    logger.debug(() -> "failed to delete file [" + existingFile + "]", ex);
                    // ignore, we don't really care, will get deleted later on
                }
```

bcully replied:
My thinking was that this same method gets called a little later, where it does do some [more complicated decision making](https://github.com/elastic/elasticsearch/blob/d6f8de22d697cd4db7a6f0eb179d6835867e9d35/server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java#L539) around exceptions. We could replicate some of that logic here, which could save some transfer work if a recovery is doomed for file system reasons, but my first instinct was to avoid adding new failure modes to recovery, and that replicating or refactoring the handling logic wasn't worth it for this case, which I wouldn't expect to be common enough to need optimization.

bcully replied:
I had actually misread this comment - I now see your point was that IllegalStateExceptions were escaping here and could fail recovery before we had a chance to pull, which might fix the problem.

I think the errors here (delete failure) are not likely to be repaired by the peer, so we might as well bail out here and save some potentially expensive transfer for a likely doomed recovery attempt. The verification step also throws some IllegalStateExceptions, which I also think are an abnormal condition (cleanup is supposed to reconcile the directory with the provided snapshot) that needs investigation.

I've taken David's suggestion and treated the directory as having an empty snapshot (which will cause cleanup to remove everything), which I think makes cleanup more robust to partially transferred indices (at the cost of potentially retransmitting some files).

I could go either way on that, I don't really have a sense for how much extra work that could create in practice.

Tim-Brooks commented on server/src/internalClusterTest/java/org/elasticsearch/recovery/TruncatedRecoveryIT.java:
```diff
@@ -102,20 +104,50 @@ public void testCancelRecoveryAndResume() throws Exception {
         final CountDownLatch latch = new CountDownLatch(1);
         final AtomicBoolean truncate = new AtomicBoolean(true);
+        IndicesService unluckyIndices = internalCluster().getInstance(IndicesService.class, unluckyNode);
         for (NodeStats dataNode : dataNodeStats) {
```
I'm pretty sure we can get ride of this and just add the send behavior from primary -> unlucky

bcully replied:
I'm not following this. I was using this to get the shard path on disk where I could scribble junk files and then check they were removed. It did seem a little indirect - is there a nicer way?

Tim-Brooks reviewed (commented):
A couple comments. I think we'll need @DaveCTurner to verify that this high level design is what he had in mind.

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into recovery-clean-before-transfer

bcully committed:
logger cleanup (PR feedback)

DaveCTurner commented on server/src/internalClusterTest/java/org/elasticsearch/recovery/TruncatedRecoveryIT.java:
```diff
@@ -102,20 +104,50 @@ public void testCancelRecoveryAndResume() throws Exception {
                         if (action.equals(PeerRecoveryTargetService.Actions.FILE_CHUNK)) {
                             RecoveryFileChunkRequest req = (RecoveryFileChunkRequest) request;
                             logger.info("file chunk [{}] lastChunk: {}", req, req.lastChunk());
                             if ((req.name().endsWith("cfs") || req.name().endsWith("fdt")) && req.lastChunk() && truncate.get()) {
```
I'm wondering why we write a garbage file only under these conditions. Could we write the garbage file always, say on receipt of the first `FILE_CHUNK` message?

bcully replied:
We could. I think we just need a little extra bookkeeping so that we create garbage once per shard rather than once per test or once per chunk. Wasn't sure it was worth it - the filter here seemed like a reasonable proxy. But I can add the explicit bookkeeping if you prefer.

DaveCTurner replied:
Yeah I'd rather we were explicit here, otherwise future-me is going to wonder what's special about these particular `FILE_CHUNK` messages.

DaveCTurner commented on server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java:
```diff
@@ -397,6 +397,32 @@ record StartRecoveryRequestToSend(StartRecoveryRequest startRecoveryRequest, Str
+                        // example.
+                        logger.info(() -> format("cleanup before peer recovery failed on shard [{}]", indexShard.shardId()), e);
+                        try {
+                            recoveryTarget.cleanTempFiles();
```
should we hold a ref to `recoveryTarget` for this call?

bcully replied:
probably yes, I'll add that.

DaveCTurner commented on server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java:
```diff
@@ -397,6 +397,32 @@ record StartRecoveryRequestToSend(StartRecoveryRequest startRecoveryRequest, Str
+                    store.incRef();
+                    try {
+                        store.cleanupAndVerify("cleanup before peer recovery", indexShard.snapshotStoreMetadata());
+                    } catch (IOException e) {
```
I think I might prefer to restrict this `catch` just to exceptions from `indexShard.snapshotStoreMetadata()` rather than from `cleanupAndVerify()` too. If `cleanupAndVerify()` fails then that seems like it maybe should be fatal?

Also could we use `org.elasticsearch.index.store.Store.MetadataSnapshot#EMPTY` to represent the snapshot if we cannot get a proper `MetadataSnapshot`? Any file re-use during recovery is going to require a valid `MetadataSnapshot`, so failing here means that the directory contents are of no further use.

bcully replied:
Tim made a similar comment, to which I replied above. There is some logic in `cleanFiles` that inspects the exception generated by the block containing `cleanupAndVerify` and does a deeper clean in some cases. Looking at it again, I think it's not just an optimization. We'd probably need to replicate that to avoid having to worry about corrupt directories that we can never make forward progress on because we never get a chance to deep clean them.

Is it always true that the files are useless if we can't get a metadata snapshot? I had been thinking that for instance we might have transferred a bunch of data files previously but not the root segment file. In that case, after transferring the segment file we could reuse the data files already transferred?

DaveCTurner replied:
> Is it always true that the files are useless if we can't get a metadata snapshot?

Yes I think so, we treat a failure as if the store is empty, see how we compute `metadataSnapshot` in `org.elasticsearch.indices.recovery.PeerRecoveryTargetService#getStartRecoveryRequest` and then use this data (alone) to compute the collection of files to recover in the implementations of `org.elasticsearch.indices.recovery.plan.RecoveryPlannerService#computeRecoveryPlan`.

I mean you're right, in principle we could work harder to determine which exact files we do/don't have and avoid a little extra data transfer sometimes. But we don't: this is already on a failure path, and not a particularly well-trodden one, so there's no need for heroic optimizations.

bcully replied:
I thought a bit more about this and am inclined to use `Store.MetadataSnapshot#EMPTY` when `snapshotStoreMetadata()` throws. Interestingly, when I tried it, I found that `store.cleanupAndVerify` appears to throw on an empty directory, [here](https://github.com/elastic/elasticsearch/blob/8efd08b019b9160f5e703520c6ce3b6a9f92cfbd/server/src/main/java/org/elasticsearch/index/store/Store.java#L699):

```
...
Caused by: java.io.FileNotFoundException: no segments* file found in store(ByteSizeCachingDirectory(ElasticsearchMockDirectoryWrapper(NIOFSDirectory@/Users/brendan/dev/elasticsearch/server/build/testrun/internalClusterTest/temp/org.elasticsearch.recovery.TruncatedRecoveryIT_31340047D7FA1434-001/tempDir-002/node_t4/indices/QuuB7D0wSki7rj5uryRXWQ/4/index lockFactory=org.apache.lucene.store.NativeFSLockFactory@b7a4d9))): files: [write.lock]
	at org.apache.lucene.index.SegmentInfos$FindSegmentsFile.run(SegmentInfos.java:801)
	at org.apache.lucene.index.SegmentInfos$FindSegmentsFile.run(SegmentInfos.java:757)
	at org.apache.lucene.index.SegmentInfos.readLatestCommit(SegmentInfos.java:535)
	at org.apache.lucene.index.SegmentInfos.readLatestCommit(SegmentInfos.java:519)
	at org.elasticsearch.common.lucene.Lucene.readSegmentInfos(Lucene.java:116)
	at org.elasticsearch.index.store.Store.readSegmentsInfo(Store.java:224)
	at org.elasticsearch.index.store.Store$MetadataSnapshot.loadFromIndexCommit(Store.java:822)
	at org.elasticsearch.index.store.Store.getMetadata(Store.java:298)
	at org.elasticsearch.index.store.Store.getMetadata(Store.java:263)
	at org.elasticsearch.index.store.Store.cleanupAndVerify(Store.java:699)
...
```

Given the name of the variable being assigned here is `metadataOrEmpty` it seems like it might be a bug, and we just haven't tested against an empty snapshot. If you agree, I'll attempt to fix that and go with this approach of falling back to `Empty`.

bcully replied:
https://github.com/elastic/elasticsearch/pull/116059 adds support for empty stores to `Store#cleanupAndVerify`

bcully replied:
after making `cleanupAndVerify` handle an empty snapshot, I've taken your suggestion and treated the directory as having an empty snapshot if we fail to load one.

DaveCTurner commented on server/src/internalClusterTest/java/org/elasticsearch/recovery/TruncatedRecoveryIT.java:
```diff
@@ -102,20 +104,50 @@ public void testCancelRecoveryAndResume() throws Exception {
                                 latch.countDown();
                                 throw new RuntimeException("Caused some truncated files for fun and profit");
                             }
+                        } else if (action.equals(PeerRecoveryTargetService.Actions.FILES_INFO) && truncate.get() == false) {
```
Do we need to check `truncate.get() == false` in this condition? I think that's there so that we only run this check on the second recovery, but sometimes we're going to get here before setting `truncate` to `false`, and there should be no garbage files on the first recovery either right?

bcully replied:
That's true. I think we're going to pass through this phase for most shards before we write the garbage into place, but it is harmless to assert that the index starts out clean. I'll remove the guard.

DaveCTurner reviewed (commented):
Trickier than I expected :) I left a few comments.

bcully committed:
PR feedback

Remove extra guard that we are in phase 2 before checking for garbage.
Take a reference on recoveryTarget around invoking cleanTempFiles.

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into recovery-clean-before-transfer

bcully committed:
Update comment

bcully committed:
PR feedback, do not intermingle garbage generation and recovery failure injection

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into recovery-clean-before-transfer

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into recovery-clean-before-transfer

bcully committed:
PR feedback: treat unreadable recovery target directory as empty

This should make pre-recovery cleanup more robust, though it may
sometimes cause some intact files to get removed and retransmitted,
if they are not part of a readable snapshot.

bcully commented:
@Tim-Brooks @DaveCTurner gentle reminder for a rereview, thanks!

bcully commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into recovery-clean-before-transfer

Tim-Brooks reviewed (approved):
LGTM

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/116854)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/116854">](https://github.com/elastic/elasticsearch/pull/116854) |
